File: webmvp/src/lib/wrapLyricLine.ts
import { getMarkStart, normalizeChordMark } from "@/lib/chordMarks";
import type { ChordMark, LyricLine } from "@/lib/types";

export type WrappedSegment = {
  lyrics: string;
  chords: ChordMark[];
};

/** Approximate monospace character width as a ratio of font size. */
export const MONO_CHAR_WIDTH_RATIO = 0.602;

export function charsPerLine(containerWidth: number, fontSize: number): number {
  if (containerWidth <= 0 || fontSize <= 0) {
    return 32;
  }
  return Math.max(12, Math.floor(containerWidth / (fontSize * MONO_CHAR_WIDTH_RATIO)));
}

function chordsForSegment(
  chords: ChordMark[],
  segStart: number,
  segEnd: number,
): ChordMark[] {
  return chords
    .filter((chord) => {
      const start = getMarkStart(chord);
      return start >= segStart && start < segEnd;
    })
    .map((chord) =>
      normalizeChordMark({
        ...chord,
        start: getMarkStart(chord) - segStart,
        end: getMarkStart(chord) - segStart + 1,
      }),
    );
}

export function wrapLyricLine(line: LyricLine, maxChars: number): WrappedSegment[] {
  const text = line.lyrics;
  const normalizedChords = line.chords.map(normalizeChordMark);

  if (maxChars <= 0 || text.length <= maxChars) {
    return [
      {
        lyrics: text,
        chords: normalizedChords,
      },
    ];
  }

  const segments: WrappedSegment[] = [];
  let offset = 0;

  while (offset < text.length) {
    const remaining = text.length - offset;
    if (remaining <= maxChars) {
      segments.push({
        lyrics: text.slice(offset),
        chords: chordsForSegment(normalizedChords, offset, text.length),
      });
      break;
    }

    let chunkLen = maxChars;
    let chunk = text.slice(offset, offset + chunkLen);

    if (offset + chunkLen < text.length && !/\s/.test(text[offset + chunkLen] ?? "")) {
      const lastSpace = chunk.lastIndexOf(" ");
      if (lastSpace > 0) {
        chunkLen = lastSpace;
        chunk = text.slice(offset, offset + chunkLen);
      }
    }

    segments.push({
      lyrics: chunk,
      chords: chordsForSegment(normalizedChords, offset, offset + chunkLen),
    });

    offset += chunkLen;
    while (offset < text.length && text[offset] === " ") {
      offset += 1;
    }
  }

  return segments.length > 0
    ? segments
    : [{ lyrics: text, chords: normalizedChords }];
}

File: mobile/lib/domain/wrap_lyric_line.dart
import 'package:lf_chords/data/models/chord_mark.dart';
import 'package:lf_chords/data/models/lyric_line.dart';
import 'package:lf_chords/domain/chord_marks.dart';

class WrappedSegment {
  const WrappedSegment({required this.lyrics, required this.chords});

  final String lyrics;
  final List<ChordMark> chords;
}

const double monoCharWidthRatio = 0.602;

int charsPerLine(double containerWidth, double fontSize) {
  if (containerWidth <= 0 || fontSize <= 0) {
    return 32;
  }
  return (containerWidth / (fontSize * monoCharWidthRatio))
      .floor()
      .clamp(12, 999999);
}

List<ChordMark> _chordsForSegment(
  List<ChordMark> chords,
  int segStart,
  int segEnd,
) {
  return chords
      .where((chord) {
        final start = getMarkStart(chord);
        return start >= segStart && start < segEnd;
      })
      .map(
        (chord) => normalizeChordMark(
          ChordMark(
            chord: chord.chord,
            start: getMarkStart(chord) - segStart,
            end: getMarkStart(chord) - segStart + 1,
          ),
        ),
      )
      .toList();
}

List<WrappedSegment> wrapLyricLine(LyricLine line, int maxChars) {
  final text = line.lyrics;
  final normalizedChords = line.chords.map(normalizeChordMark).toList();

  if (maxChars <= 0 || text.length <= maxChars) {
    return [
      WrappedSegment(lyrics: text, chords: normalizedChords),
    ];
  }

  final segments = <WrappedSegment>[];
  var offset = 0;

  while (offset < text.length) {
    final remaining = text.length - offset;
    if (remaining <= maxChars) {
      segments.add(
        WrappedSegment(
          lyrics: text.substring(offset),
          chords: _chordsForSegment(normalizedChords, offset, text.length),
        ),
      );
      break;
    }

    var chunkLen = maxChars;
    var chunk = text.substring(offset, offset + chunkLen);

    if (offset + chunkLen < text.length &&
        !RegExp(r'\s').hasMatch(text[offset + chunkLen])) {
      final lastSpace = chunk.lastIndexOf(' ');
      if (lastSpace > 0) {
        chunkLen = lastSpace;
        chunk = text.substring(offset, offset + chunkLen);
      }
    }

    segments.add(
      WrappedSegment(
        lyrics: chunk,
        chords: _chordsForSegment(normalizedChords, offset, offset + chunkLen),
      ),
    );

    offset += chunkLen;
    while (offset < text.length && text[offset] == ' ') {
      offset += 1;
    }
  }

  return segments.isNotEmpty
      ? segments
      : [WrappedSegment(lyrics: text, chords: normalizedChords)];
}
